safehttp/response_writer.go

```diff
@@ -26,17 +26,22 @@ type ResponseWriter struct {
 	// Having this field unexported is essential for security. Otherwise one can
 	// easily overwrite the struct bypassing all our safety guarantees.
 	header  Header
+	handler handlerWithInterceptors
+	req     *IncomingRequest
 	written bool
 }
 
 // NewResponseWriter creates a ResponseWriter from a safehttp.Dispatcher, an
-// http.ResponseWriter and a list of interceptors associated with a ServeMux.
-func NewResponseWriter(d Dispatcher, rw http.ResponseWriter) *ResponseWriter {
+// http.ResponseWriter and a reference to the current IncomingRequest being served.
+// The IncomingRequest will only be used by the commit phase, which only runs when
+// the ServeMux is used, and can be passed as nil in tests.
+func NewResponseWriter(d Dispatcher, rw http.ResponseWriter, req *IncomingRequest) *ResponseWriter {
```

Comment:
Let's finish this PR, but the `IncomingRequest` should not be added here, please leave a TODO.